### PR TITLE
Api cleanup (remove List and Dictionary in favor of IList and IDictionar...

### PR DESCRIPTION
**Pull request by yishaigalatzer**
Api cleanup (remove List and Dictionary in favor of IList and IDictionary)
from the public surface copy header values on registration.

Make a copy of the mediatypeheader value in FormatterMappings so it's
really immutable.

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Core/ActionDescriptor.cs [hunk 1]
```diff
@@ -17,20 +17,20 @@ public ActionDescriptor()
 
         public virtual string Name { get; set; }
 
-        public List<RouteDataActionConstraint> RouteConstraints { get; set; }
+        public IList<RouteDataActionConstraint> RouteConstraints { get; set; }
 
         public AttributeRouteInfo AttributeRouteInfo { get; set; }
 
-        public Dictionary<string, object> RouteValueDefaults { get; private set; }
+        public IDictionary<string, object> RouteValueDefaults { get; }
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ActionDescriptor.cs]: This seems like it's going to break stuff

File: src/Microsoft.AspNet.Mvc.Core/Description/ApiDescriptionProviderContext.cs [hunk 2]
```diff
@@ -29,6 +29,6 @@ public ApiDescriptionProviderContext([NotNull] IReadOnlyList<ActionDescriptor> a
         /// <summary>
         /// The list of resulting <see cref="ApiDescription"/>.
         /// </summary>
-        public List<ApiDescription> Results { get; private set; }
+        public IList<ApiDescription> Results { get; private set; }
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Description/ApiDescriptionProviderContext.cs]: can probably remove the provider set here

File: src/Microsoft.AspNet.Mvc.Core/Description/ApiDescription.cs [hunk 3]
```diff
@@ -40,7 +40,7 @@ public ApiDescription()
         /// <summary>
         /// The list of <see cref="ApiParameterDescription"/> for this api.
         /// </summary>
-        public List<ApiParameterDescription> ParameterDescriptions { get; private set; }
+        public IList<ApiParameterDescription> ParameterDescriptions { get; private set; }
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Description/ApiDescription.cs]: can probably remove the provider set here

**Comment by rynowak** [comment]
:shipit:

File: src/Microsoft.AspNet.Mvc.Core/Description/ApiDescriptionProviderContext.cs (hunk 2 above)
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/Description/ApiDescriptionProviderContext.cs]: will follow up in a separate cleanup

**Comment by yishaigalatzer** [comment]
Merged

**Pull request closed without merging**